pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go

```diff
@@ -0,0 +1,183 @@
+	// first-class user entry on the LDAP server, given a first-class group entry with member records
+	// stored in an attribute.
+	userQueryTemplate api.LDAPQueryFromAttributeTemplate
+	// groupNameAttribute is the list of LDAP group entry attributes that correspond to the OpenShift
+	// Group Name field.
+	groupNameAttribute []string
+	// LDAPUserAttributeDefiner defines the values corresponding to OpenShift Identities in LDAP entries
+	// by using a deterministic mapping of LDAP entry attributes to OpenShift Identity fields
+	userAttributeDefiner ldaputil.LDAPUserAttributeDefiner
+
+	// cachedGroups holds the result of group queries for later reference, indexed on group UID
+	// e.g. this will map an LDAP group UID to the LDAP entry returned from the query made using it
+	cachedGroups map[string]*ldap.Entry
+	// cachedUsers holds the result of user queries for later reference, indexed on user UID
+	// e.g. this will map an LDAP user UID to the LDAP entry returned from the query made using it
+	cachedUsers map[string]*ldap.Entry
+}
+
+// ExtractMembers returns the LDAP member entries for a group specified with a ldapGroupUID
+func (e *LDAPInterface) ExtractMembers(ldapGroupUID string) (members []*ldap.Entry, err error) {
+	// get group entry from LDAP
+	group, err := e.GroupEntryFor(ldapGroupUID)
+	if err != nil {
+		return nil, err
+	}
+
+	// extract member UIDs from group entry
+	ldapMemberUIDs := group.GetAttributeValues(e.userQueryTemplate.SpecificFilterSourceAttribute)
+
+	// find members on LDAP server or in cache
+	for _, ldapMemberUID := range ldapMemberUIDs {
+		memberEntry, err := e.userEntryFor(ldapMemberUID)
+		if err != nil {
+			return nil, err
+		}
+		members = append(members, memberEntry)
+	}
+	return members, nil
+}
+
+// GroupFor returns an LDAP group entry for the given group UID by searching the internal cache
+// of the LDAPInterface first, then sending an LDAP query if the cache did not contain the entry.
+// This also satisfies the LDAPGroupGetter interface
+func (e *LDAPInterface) GroupEntryFor(ldapGroupUID string) (group *ldap.Entry, err error) {
+	group, exists := e.cachedGroups[ldapGroupUID]
+	if !exists {
+		group, err = e.queryForGroup(ldapGroupUID)
+		if err != nil {
+			return nil, err
+		}
+		// cache for annotation extraction
+		e.cachedGroups[ldapGroupUID] = group
+	}
+	return group, nil
+}
+
+// queryForGroup queries for a specific group identified by a ldapGroupUID with the query config stored
+// in a LDAPInterface
+func (e *LDAPInterface) queryForGroup(ldapGroupUID string) (group *ldap.Entry, err error) {
+	// determine list of attributes to request from LDAP server
+	attributes := util.NewStringSet(e.groupNameAttribute...)
+	attributes.Insert(e.userQueryTemplate.SpecificFilterSourceAttribute)
+
+	// build specific filter by injecting ldapGroupUID
+	specificFilter := fmt.Sprintf(e.groupQueryTemplate.SpecificFilterBase, ldapGroupUID)
```

Comment:
What if ldapGroupUid is a DN? How would they construct the SpecificFilterBase to query for it